cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataUpgrader.java

```diff
@@ -51,130 +46,114 @@
-  private final NamespaceQueryAdmin namespaceQueryAdmin;
-
-  // A single thread executor to perform all upgrade related work
-  private ExecutorService executor;
 
   @Inject
   MetadataUpgrader(DatasetInstanceManager datasetInstanceManager,
-                   ExistingEntitySystemMetadataWriter existingEntitySystemMetadataWriter, MetadataStore metadataStore,
-                   DatasetFramework datasetFramework, NamespaceQueryAdmin namespaceQueryAdmin) {
+                   ExistingEntitySystemMetadataWriter existingEntitySystemMetadataWriter,
+                   MetadataStore metadataStore, DatasetFramework datasetFramework) {
     this.existingEntitySystemMetadataWriter = existingEntitySystemMetadataWriter;
     this.metadataStore = metadataStore;
     this.datasetFramework = datasetFramework;
-    this.namespaceQueryAdmin = namespaceQueryAdmin;
   }
 
-  public void executeUpgrade() {
+  public void createOrUpgradeIfNecessary() {
     // Upgrade the Specs inline
-    upgradeMetadataDatasetSpecs();
+    upgradeMetadataDatasetSpecs(MetadataScope.SYSTEM);
+    upgradeMetadataDatasetSpecs(MetadataScope.USER);
 
     // Create a single thread executor using daemon threads which runs data-migration process.
-    executor = Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory("metadata-upgrader"));
+    ExecutorService executor =
+      Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory("metadata-upgrader"));
     executor.submit(new Callable<Void>() {
       @Override
       public Void call() throws Exception {
-        runUpgrade();
+        runSystemUpgrade();
+        runUserUpgrade();
         return null;
       }
     });
   }
 
-  private void runUpgrade() throws Exception {
-    Predicate<Throwable> alwaysRetry = new Predicate<Throwable>() {
-      @Override
-      public boolean apply(@Nullable Throwable input) {
-        return true;
-      }
-    };
+  private boolean isUpgradeRequired(final MetadataScope scope) throws Exception {
     boolean upgradeRequired;
     try {
       upgradeRequired = Retries.callWithRetries(new Retries.Callable<Boolean, Exception>() {
         @Override
         public Boolean call() throws Exception {
-          return (metadataStore.isUpgradeRequired(MetadataScope.USER)
-            || metadataStore.isUpgradeRequired(MetadataScope.SYSTEM));
+          return metadataStore.isUpgradeRequired(scope);
         }
       }, retryStrategy);
     } catch (Exception e) {
-      LOG.error("Failed to check Metadata Dataset Upgrade status.", e);
+      LOG.error("Failed to check Metadata Dataset Upgrade status for scope {}.", scope, e);
       throw e;
     }
+    return true; //upgradeRequired;
+  }
 
-    if (!upgradeRequired) {
-      LOG.info("Metadata Dataset Upgrade is not required.");
-      return;
-    }
-
-    LOG.info("Metadata DataSet Upgrade is reqiured.");
+  private void saveUpgradeState(final MetadataScope scope) throws Exception {
```

Comment:
Since the end result is actually a tag with the current version and this is also called when getting created the first time, saveUpgradeState seemed reasonable. I have renamed it to markUpgradeComplete() anyway.